File: SALARY/salary.rb
#!/usr/bin/env ruby

class Equalizer
	def initialize(nworkers, salaries)		
		raise 'Error: number of workers is different from number of salaries' unless nworkers == salaries.length	
		@salaries = salaries
	end

	def minimum_operations
		number_of_iterations = 0
		while (@salaries.count(@salaries[0]) != @salaries.length) do 
			max_index = @salaries.index(@salaries.max)
			@salaries.each_with_index do |salary, index|
				@salaries[index] += 1 unless max_index == index
			end		
			number_of_iterations += 1
		end

		number_of_iterations
	end
end

class Reader
	def initialize(file)
		@file = file
	end

	def read()
		equalizers = []

		lines = @file.readlines 
		size = lines[0].to_i

		lines[1..lines.length].each_slice(2) do |slice|
			nworkers = slice[0].chomp.to_i
			salaries = slice[1].chomp.split(' ').map { |x| x.to_i }
			equalizers << Equalizer.new(nworkers, salaries)
		end

		equalizers	
	end
end

equalizers = Reader.new(ARGF).read()

equalizers.each do |eq|
	puts eq.minimum_operations
end

File: SALARY/salary.cpp

#include <stdio.h>

int SALARIES[100];

bool all_eq(int nworkers) {
	int previous = SALARIES[0];
	
	for(int i = 1; i < nworkers; i++){

		if(previous != SALARIES[i]) {
			return false;
		}

		previous = SALARIES[i];
	}

	return true;
}

int index_of_max_value(int nworkers){
	int max = SALARIES[0];
	int max_index = 0;
	
	for(int i = 1; i < nworkers; i++){
		if(max < SALARIES[i]) {
			max = SALARIES[i];
			max_index = i;
		}
	}

	return max_index;
}

int main() {

	int test_cases;
	int nworkers;
	int max_index;
	int i,j;
	int noperations;

	// first line is the number of test_cases
	scanf("%d\n", &test_cases);

	// each test case
	for(i = 0; i < test_cases; i++){

		// reading initial salaries
		scanf("%d\n", &nworkers);
		for(j = 0; j < nworkers; j++){
			scanf("%d ", &SALARIES[j]);			
		}

		noperations = 0;
		while(!all_eq(nworkers)){
			noperations++;
			max_index = index_of_max_value(nworkers);
			for(j = 0; j < nworkers; j++){
				SALARIES[j]++;
			}
			SALARIES[max_index]--;
		}

		printf("%d\n", noperations);
	}
}

